test/extensions/filters/http/ext_authz/config_test.cc

```diff
@@ -1,97 +1,201 @@
+        std::cerr << "worker!\n";
+      });
+    }
+
+    // Notify all the worker to continue.
+    workers_should_fire_.Notify();
+  }
+
+protected:
+  void startThreading() {
+    spawnMain();
+    spawnWorkerThreads();
+  }
+
+  void cleanUpThreading() {
+    tls_.shutdownGlobalThreading();
+    std::cerr << "start to join workers!\n";
+    joinWorkers();
+    std::cerr << "join workers complete!\n";
+    joinMain();
+    std::cerr << "join main complete!\n";
+  }
+
+  ThreadLocal::InstanceImpl& tls() { return tls_; }
+
+  Api::Api& api() { return *api_; }
+
+private:
+  void spawnMain() {
+    main_dispatcher_ = api_->allocateDispatcher("main");
+    main_thread_ = api_->threadFactory().createThread(
+        [this]() { main_dispatcher_->run(Event::Dispatcher::RunType::RunUntilExit); });
+    tls().registerThread(*main_dispatcher_, true);
+  }
+  void spawnWorkerThreads() {
+    absl::BlockingCounter start_counter(num_threads_);
+    // Create worker dispatchers and register tls.
+    for (size_t i = 0; i < num_threads_; i++) {
+      worker_dispatchers_.emplace_back(api_->allocateDispatcher(std::to_string(i)));
+    }
+    for (size_t i = 0; i < num_threads_; i++) {
+      tls().registerThread(*worker_dispatchers_[i], false);
+      // i must be explicitly captured by value.
+      worker_threads_.emplace_back(api_->threadFactory().createThread([this, i, &start_counter]() {
+        start_counter.DecrementCount();
+        worker_dispatchers_[i]->run(Event::Dispatcher::RunType::RunUntilExit);
+      }));
+    }
+    start_counter.Wait();
+  }
+
+  void joinWorkers() {
+    for (Event::DispatcherPtr& dispatcher : worker_dispatchers_) {
+      dispatcher->exit();
+      std::cerr << "exit\n";
+    }
+    for (Thread::ThreadPtr& worker : worker_threads_) {
+      worker->join();
+      std::cerr << "worker join\n";
+    }
+  }
+  void joinMain() {
+    main_dispatcher_->exit();
+    main_thread_->join();
+  }
+
+  const size_t num_threads_;
+
+  ThreadLocal::InstanceImpl tls_;
+  Api::ApiPtr api_;
 
-TEST(HttpExtAuthzConfigTest, CorrectProtoHttp) {
-  std::string yaml = R"EOF(
+  std::vector<Event::DispatcherPtr> worker_dispatchers_;
+  std::vector<Thread::ThreadPtr> worker_threads_;
+
+  Event::DispatcherPtr main_dispatcher_;
+  Thread::ThreadPtr main_thread_;
+
+  absl::Notification workers_should_fire_;
+};
+
+class ExtAuthzFilterTest : public MultiThreadTest, public testing::Test {
+public:
+  ExtAuthzFilterTest() : MultiThreadTest(10), stat_names_(symbol_table_) {}
+
+  Http::FilterFactoryCb createFilterFactory(const std::string& ext_authz_config_yaml) {
+    async_client_manager_ = std::make_unique<TestAsyncClientManagerImpl>(
+        context_.cluster_manager_, tls(), connection_.stream_info_.ts_, api(), stat_names_);
+    EXPECT_CALL(context_, getServerFactoryContext())
+        .WillRepeatedly(testing::ReturnRef(server_context_));
+    ON_CALL(context_.cluster_manager_.async_client_manager_, getOrCreateRawAsyncClient(_, _, _, _))
```

Comment:
I think you will see TSAN errors in this expectation, since gmock's internal bookkeeping structures for tracking expectations are not thread safe.
I think you will need to actually create the async_client_manager that is not mocked.

Reply:
We wouldn't have a race here if ON_CALL don't modify the state of the object.